Brackets based parser (fixes #667)

Playacem opened the pull request:
This is a brackets based parser as described in #667. 

Currently the new parser is not supported in the OpenApi DSL. 
I added a small abstraction layer to the path parser creation. The interface only contains the public methods for now.

# Implementation details

- Brackets are forbidden as part of a parameter name but wildcards are allowed
- Brackets are counted for each segment separately to provide an early way out
- Wildcards cannot be directly adjacent to a path parameter. Otherwise `/*{hi}` would behave differently from `/{hi}*`. In the second case the path parameter would only accept a single character.

I am unsure on how to adapt the OpenApiDsl to the new bracket parser. Might extend the interface with the internal variables required. Some parts would need to be adopted to deal with the new MultipleSegments class in PathSegment2.

I look forward to feedback.

Playacem commented:
The OpenApi plugin and the RedirectToLowerCasePathPlugin now use the new abstraction layer to interact with the `PathParser`.

Tests for these two plugins using the brackets based parser are still missing at the moment.

tipsy commented:
Thanks @Playacem, nice job, it looks like a lot of work has gone into this. I haven't read through all of the code, but I have a high level comment.

I don't think we should support two different ways of doing things, so if we introduce delimiters, we should get rid of the colon based parser. I supposed we could make it possible for the parser to work with only an opening delimiter to support the old syntax, but I don't really see much value in it.

Playacem commented:
Thanks for the feedback @tipsy. 

> I don't think we should support two different ways of doing things, so if we introduce delimiters, we should get rid of the colon based parser.

I designed the implementation in such a way as to not break anything. I can make a pull request which completely replaces the colon based parser instead. That way I can remove the indirection I created with `PathParserSpec`. 

> I supposed we could make it possible for the parser to work with only an opening delimiter to support the old syntax, but I don't really see much value in it.

Something like `{hello` would just look kinda weird but shoudn't be too hard to implement.

> Basically, I want this for Javalin 4 rather than 3.X, but if you want it in 3.X I think we should make one parser that supports using only opening delimiters. This parser would have the delimiter char set to : by default, but people could change it to {}.

I personally can wait until Javalin 4. I assume you want to have it in 4.0 as a breaking change?

Had the idea to introduce a different type of path parameter that ignores slashes for those that miss the `splat` feature of Javalin 2.

Example: Path `hello/with/slashes` gets matched by `hello/[input]` and `input` has a value of `with/slashes`.

tipsy commented:
> Something like `/{hello` would just look kinda weird but shouldn't be too hard to implement.

It would look weird with `{`, but if you also make the delimiter characters configurable you can use `:`, which would make the change backwards compatible.

> I personally can wait until Javalin 4. I assume you want to have it in 4.0 as a breaking change?

Yes, unless it's easy to support 1 delimiter + configurable delimiter characters. My main concern is making the code-base too complex, this feature shouldn't require +400 lines.

> Had the idea to introduce a different type of path parameter that ignores slashes for those that miss the splat feature of Javalin 2.

I think that would be good, I underestimated the number of people who were using this feature when I removed it.

Playacem commented:
Made some changes to the new PathParser.
It can now completely replace the old parser and splat like brackets have been added.
I did leave in the new abstraction layer as this allows the separation of configuration (`JavalinPathParser`) of a PathParser with it's creation (`createPathParser`).

Not really supported (yet) is the behaviour of optional ending in combination with multiple segments, e.g. `hi-:name`.

tipsy commented:
@Playacem this might be relevant again soon :)

Playacem commented:
Probably have to do some work to base this pr onto the latest changes. The idea of using a state machine for parsing the path segments still seems like a good idea.

I can probably remove a lot of code that is only there for backwards compatibilty. 
Will add a new path parser to the 4.0 wishlist :)

Playacem commented:
The remaining three broken tests all involve wildcards at the end of a path. Caused by the support for non sub path wildcards [added in November 2020.](https://github.com/tipsy/javalin/commit/6855017c267ad258c534765249d4017032af96aa)

Currently looking into for ideas on how to solve this. 
Possible options:
- rewrite failing tests -> at least two of them represent actual functionality of the new parser
- integrate the feature into the parser itself?

Playacem commented:
@tipsy just saw that you added the feature that is causing me problems. 
Can you give me more context on why you made this change? https://github.com/tipsy/javalin/commit/6855017c267ad258c534765249d4017032af96aa

tipsy commented:
@Playacem sorry, I didn't see your comment! 

I found myself repeating `before` filters two times for every `path` group.

![image](https://user-images.githubusercontent.com/1521451/111834845-761de300-88f4-11eb-90b2-a23831359c54.png)

```
before("/my-path")
path("/my-path") {
    before("/my-path/")
}
```

can now be written

```
before("/my-path*")
path("/my-path") {
}
```

tipsy commented:
@Playacem do you think you'll look at this anytime soon?

Playacem commented:
> @Playacem do you think you'll look at this anytime soon?

I plan to look into this next weekend.
My current idea for the `/path*` feature is to replace it with the slash accepting brackets `/path{*}`.

tipsy commented:
Great, looking forward to it!

Playacem commented:
Well, now I broke all MicroMeter tests only on Linux. That is a first for me. :D
Will look into it tomorrow or next weekend. Will have to figure out a local setup to debug this.

The non sub-path wildcards are integrated as `/{path}**` by creating multiple regular expressions if detected.

Playacem commented:
Cannot reproduce the issue by running the test suite under WSL. Apart from setting up a real VM I am out of ideas on this one :/

Playacem commented:
The MicroMeter plugin tests fail at [Javalin.java line 181](https://github.com/Playacem/javalin/blob/130d250fbd95b0ce5744bd2c53ca61f48ef2ec87/javalin/src/main/java/io/javalin/Javalin.java#L181)
Here are the lines in question:
```java
        try {
            Javalin.log.info("Starting Javalin ...");
            server.start(wsServlet);
            Javalin.log.info("Javalin started in " + (System.currentTimeMillis() - startupTimer) + "ms \\o/");
            eventManager.fireEvent(JavalinEvent.SERVER_STARTED);
        } catch (Exception e) {
            Javalin.log.error("Failed to start Javalin");
            eventManager.fireEvent(JavalinEvent.SERVER_START_FAILED);
```

Javalin.log seems to be null for some unknown reason. This seems to be the actual reason as well, as as far as I can tell the `Starting Javalin` log line never appears. Truly a mistery...

tipsy commented:
> Javalin.log seems to be null for some unknown reason. This seems to be the actual reason as well, as as far as I can tell the Starting Javalin log line never appears. Truly a mistery...

Argh, that part is a bit hacky... I can fix it on the master branch :)

tipsy commented:
@Playacem if you rebase onto master now your issue should be gone :)

Playacem commented:
@tipsy getting some Kotlin 1.4 / 1.5 mismatch errors ([Example](https://github.com/tipsy/javalin/pull/831/checks?check_run_id=2649659241#step:5:54)). That might be something that needs fixing in the master branch.

The windows Java 11 failure is related to the Java Setup step, which might just be a fluke.

tipsy commented:
Seems it was just a fluke. I'll take a look at this tomorrow!

tipsy commented:
I've read through it now @Playacem. From looking at the test, the matching seems to be much more versatile now, which is great, but I have two comments:

1. We talked about removing the colon-based parser since we're on Javalin 4 now, I still think that's the best way to go - supporting two different parsers is more work for us, and confusing for end users.
2. Could you quickly outline the roles of `<>` `{}`,`*` and `**` ? That's a lot of symbols to keep track of. Do they behave differently in the different parsers? (if so, that's a strong argument for **1**)

Playacem commented:
> 1. We talked about removing the colon-based parser since we're on Javalin 4 now, I still think that's the best way to go - supporting two different parsers is more work for us, and confusing for end users.

This iis on my to do list. The current implementation does not use the old path parser at all. I created an compatibility mode for the new parser to behave like the old parser. 
My current plan is to remove this compatibility mode, leaving only a single parser with brackets as a result.

> 2. Could you quickly outline the roles of `<>` `{}`,`*` and `**` ? That's a lot of symbols to keep track of. Do they behave differently in the different parsers? (if so, that's a strong argument for **1**)

`*` is the same old wildcard as before.
`{}` and `<>` are the indicators for the new path parameters. `{name}` behaves like the old `:name` syntax.`<>` are path parameters that also accept slashes, mainly intended as a better solution for splats. 
`**` is the replacement for the old `:name*` syntax. When detected at the end of a path segment the parser will behave as described in the current non sub-path wildcards segment. It creates internally two sets of parsed segments one with a wildcard at the end and one without.

In general you can combine static content, path parameters and wildcards all in the same segment. `/hi-{name}-*` is a valid path segment.

Todos:
- [ ] Rebase to current master, fixing merge conflicts on the way
- [ ] Remove old `PathParser` file
- [ ] Rename `PathParser2 ` to `PathParser`
- [ ] Rename `PathSegment2` to `PathSegment`
- [ ] Remove compatibility mode for old colon parser, rewriting all tests in the process
- [ ] Update examples
- [ ] Update documentation

The removal of the compatibility mode could be done at a later point in time but I think it is best idea to rip of the band aid and convert all the tests at once.

tipsy commented:
Thanks for the thorough explanation @Playacem !

> ** is the replacement for the old :name*

Let's see if I understand this correctly: 

`/hi-{name}*` would match `/hi-placyacemTest` but not `/hi-placyacem`, while
`/hi-{name}**` would match `/hi-placyacemTest`  and also `/hi-placyacem`

And I guess it works the same for non-path-params:

`/hi-playa*` would match `/hi-playaTest` but not `/hi-playa`, while
`/hi-playa**` would match `/hi-playaTest`  and also `/hi-playa`

Is that correct?

Playacem commented:
`/hi-{name}**` is equivalent to registering two paths `/hi-{name}` and `/hi-{name}/*`. 
The same applies for non path-params. `/hi-playa**` results in `/hi-playa` and `/hi-playa/*`
`/hi-{name}*` should be illegal as the path parameter would only match a single character. Afaik it doesn't result in an exception at the moment.

tipsy commented:
Got it! I'm a bit unsure of the syntax of this, maybe we should outline all the use cases to see if we can make it more coherent? 

I think your todo list looks good, I agree that it's better to rip if the band aid. I will have a look at path matching use cases after I'm done looking at the Validation API.

Playacem commented:
Maybe we could use a different character for this behaviour. Maybe `%` or `$` but I kinda like the reference to the regular wildcard 
character.

tipsy commented:
So this is the current proposal: (?)

| Path | Matches |
| ---- | ------- |
| `/base-*` | all paths that start with `base-`, but not `base-` itself  |
| `/base-**`| all paths that start with `base-`, and also `base-` itself |
| `/base-<param>`| all paths that start with `base-`, param extractable, param can contain slashes |
| `/base-{param}`| all paths that start with `base-`, param extractable, param can't contain slashes |
| `/base-<param>-end`| all paths that start with `base-` and end with `-end`, param extractable, param can contain slashes |
| `/base-{param}-end`| all paths that start with `base-`, and end with `-end` param extractable, param can't contain slashes |

I definitely like `*` for wildcard, and I think I also like the `**` for wildcard+base.

I'm not loving the character of `<>` for path params with slashes. I guess you chose it because it's `()` and `[]` look too similar to `{}` ? `/base-*param*`  look a bit ugly, but is distinctly different and hints at the similarity to wildcards. But then `/base-<param>**` would become `/base-*param***`, which is just ridiculous 😄 

I don't really have any better suggestions from what you've made here.

I think we should also remove the splat method, since this is now replaced by `<>`.

Playacem commented:
> So this is the current proposal: (?)
> 
> Path	Matches
> `/base-*`	all paths that start with `base-`, but not `base-` itself
> `/base-**`	all paths that start with `base-`, and also `base-` itself
> `/base-<param>`	all paths that start with `base-`, param extractable, param can contain slashes
> `/base-{param}`	all paths that start with `base-`, param extractable, param can't contain slashes
> `/base-<param>-end`	all paths that start with `base-` and end with `-end`, param extractable, param can contain slashes
> `/base-{param}-end`	all paths that start with `base-`, and end with `-end` param extractable, param can't contain slashes

Added some assertions to check against the actual implementation. The `/base-*` does not behave as described in the table but otherwise it is quite on the mark. 
The double star stuff is only necessary for `/{param}**` cases if I recall correctly, as otherwise they would be infering with each other.

> I'm not loving the character of `<>` for path params with slashes. I guess you chose it because it's `()` and `[]` look too similar to `{}` ? `/base-*param*` look a bit ugly, but is distinctly different and hints at the similarity to wildcards. But then `/base-<param>**` would become `/base-*param***`, which is just ridiculous 😄
> 
Exactly!


> I think we should also remove the splat method, since this is now replaced by `<>`.
Will do so soon.

tipsy commented:
> Added some assertions to check against the actual implementation. The /base-* does not behave as described in the table but otherwise it is quite on the mark.

Any reason why it shouldn't?

Also, is there a check throwing illegalargumentexception if you do `/{param}/<param>` ?

Playacem commented:
> > Added some assertions to check against the actual implementation. The /base-* does not behave as described in the table but otherwise it is quite on the mark.
> 
> Any reason why it shouldn't?

The wildcard matches nothing as well, leading to the acceptance of `/base-`. 
We might be getting philosophical here: Should something that matches everything also match nothing? ;)
Currently for a requested path of `/base-` both `/base-*` and `/base-**` would match as routes.

> Also, is there a check throwing illegalargumentexception if you do `/{param}/<param>` ?

Not yet, will add it to my todo list :)

tipsy commented:
Sorry, I created another small conflict 😬 
I guess your test class should completely replace the `TestRouting` class? 

Do you think you'll be ready to merge soon? I would love to get this out as part of the next alpha.

Playacem commented:
I have been quite busy at work but I aim to make some progress on thursday.

Playacem commented:
> Also, is there a check throwing illegalargumentexception if you do `/{param}/<param>` ?

A basic check has been added. 
I think everything should be ready to merge now and I am looking forward to feedback :) @tipsy

tipsy commented:
Great @Playacem, I will look through it later today !

tipsy commented:
Overall it looks very good @Playacem! I've checked it out locally, and I'm wondering if we should organize the code a little differently. Currently we have 

```
io/javalin/core/JavalinPathParser.kt    (30 lines)
io/javalin/core/PathParser.kt           (295 lines)
io/javalin/http/PathMatcher.kt          (41 lines)
io/javalin/websocket/WsPathMatcher.kt   (48 lines)
```

* Can you create a `routing` package and split out some of the classes in `PathParser.kt` into separate files?
* Why make `PathParserSpec` a single implementation interface? Could we get rid of it?
* Can `INSIDE_BRACKET_TYPE_1` get a more descriptive name (`INSIDE_CURLY_BRACES`, or `INSIDE_PATH_PARAM`?)
* Can you remove `splatList`/`splat()` from `Context`?

Playacem commented:
> * Can you create a `routing` package and split out some of the classes in `PathParser.kt` into separate files?

Sure. I can move basically everything that is not inside the PathParser. Might leave the ParserState still in there as it just part of the state machine.

> * Why make `PathParserSpec` a single implementation interface? Could we get rid of it?

I originally introduced it to abstract the api between the two path parsers, now that we removed the path parser it is not really necessary anymore. It offers a nice boundary for the OpenApi stuff as that is creating their own path parser instances. With the interface and the factory method `createPathParser` I was able to remove the direct constructor calls there. 

Will remove it as the last commit switching everything back to direct contructor calls.

> * Can `INSIDE_BRACKET_TYPE_1` get a more descriptive name (`INSIDE_CURLY_BRACES`, or `INSIDE_PATH_PARAM`?)

Will go with `INSIDE_CURLY_BRACKETS` and `INSIDE_ANGLE_BRACKETS`

> * Can you remove `splatList`/`splat()` from `Context`?

Will do, together with the splat methods in the path parser.

My current plan is to do it tomorrow or on saturday.

tipsy commented:
> My current plan is to do it tomorrow or on saturday.

Sounds good, thank you for your work on this :)

Playacem commented:
I moved basically everything that did not absolutely have to be inside the PathParser class into separate files, keeping only the member methods and properties.
I also made every parser exception a sub class of `IllegalArgumentException` as it fits better with the exception purpose.

tipsy commented:
@Playacem It looks great now ! Do you still want to wait until the weekend, or are you ready for merge?

@rbygrave this change may or may not be interesting to you - I know you already changed routing in jex, but this adds some nice concepts IMO :)

rbygrave commented:
Cheers @tipsy Jex uses bracket based parsing (with regex support to match jax-rs).  I'll have a closer look at this later, I don't understand why `<`  `>` are in here (for openapi support? 2 ways to do the same thing?). Thanks.

tipsy commented:
One allows slashes inside, the other one does not.

tipsy commented:
Thanks again @Playacem !
